go/vt/mysqlctl/xtrabackupengine.go

```diff
@@ -0,0 +1,484 @@
+	}
+	pos, err := conn.MasterPosition()
+	if err != nil {
+		return false, vterrors.Wrap(err, "unable to obtain master position")
+	}
+	flavor := pos.GTIDSet.Flavor()
+	logger.Infof("Detected MySQL flavor: %v", flavor)
+
+	backupProgram := path.Join(*xtrabackupEnginePath, xtrabackup)
+
+	flagsToExec := []string{"--defaults-file=" + cnf.path,
+		"--backup",
+		"--socket=" + cnf.SocketFile,
+		"--slave-info",
+		"--user=" + *xtrabackupUser,
+		"--target-dir=" + cnf.TmpDir,
+	}
+	if *xtrabackupStreamMode != "" {
+		flagsToExec = append(flagsToExec, "--stream="+*xtrabackupStreamMode)
+	}
+
+	if *xtrabackupBackupFlags != "" {
+		flagsToExec = append(flagsToExec, strings.Fields(*xtrabackupBackupFlags)...)
+	}
+
+	backupFileName := be.backupFileName()
+
+	wc, err := bh.AddFile(ctx, backupFileName, 0)
+	if err != nil {
+		return false, vterrors.Wrapf(err, "cannot create backup file %v", backupFileName)
+	}
+	closeFile := func(wc io.WriteCloser, fileName string) {
+		if closeErr := wc.Close(); err == nil {
+			err = closeErr
+		} else if closeErr != nil {
+			// since we already have an error just log this
+			logger.Errorf("Error closing file %v", fileName)
+		}
+	}
+	defer closeFile(wc, backupFileName)
+
+	backupCmd := exec.Command(backupProgram, flagsToExec...)
+	backupOut, _ := backupCmd.StdoutPipe()
+	backupErr, _ := backupCmd.StderrPipe()
+	dst := bufio.NewWriterSize(wc, 2*1024*1024)
+	writer := io.MultiWriter(dst)
+
+	// Create the gzip compression pipe, if necessary.
+	var gzip *pgzip.Writer
+	if *backupStorageCompress {
+		gzip, err = pgzip.NewWriterLevel(writer, pgzip.BestSpeed)
+		if err != nil {
+			return false, vterrors.Wrap(err, "cannot create gziper")
+		}
+		gzip.SetConcurrency(*backupCompressBlockSize, *backupCompressBlocks)
+		writer = gzip
+	}
+
+	if err = backupCmd.Start(); err != nil {
+		return false, vterrors.Wrap(err, "unable to start backup")
+	}
+
+	// Copy from the stream output to destination file (optional gzip)
+	_, err = io.Copy(writer, backupOut)
+	if err != nil {
+		return false, vterrors.Wrap(err, "cannot copy output from xtrabackup command")
+	}
+
+	// Close gzip to flush it, after that all data is sent to writer.
+	if gzip != nil {
+		if err = gzip.Close(); err != nil {
+			return false, vterrors.Wrap(err, "cannot close gzip")
+		}
+	}
+
+	// Flush the buffer to finish writing on destination.
+	if err = dst.Flush(); err != nil {
+		return false, vterrors.Wrapf(err, "cannot flush destination: %v", backupFileName)
+	}
+
+	errOutput, err := ioutil.ReadAll(backupErr)
+	backupCmd.Wait()
+	output := string(errOutput)
+
+	logger.Infof("Xtrabackup backup command output: %v", output)
+	// check for success message : xtrabackup: completed OK!
+	usable := (err == nil && strings.Contains(output, successMsg))
+	substrs := strings.Split(output, "'")
+	index := -1
+	for i, str := range substrs {
+		if strings.Contains(str, "GTID of the last change") {
```

Comment:
similar q about detection with strings

Reply:
this one, there is no other way. xtrabackup puts the GTID in a file which is bundled up inside the archive that is the backup file.